## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-zoned.sh | :x: [more...](#5099737a-4001-4518-8990-ea96bb8fb94e) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#2175e2f2-f622-416b-b468-208270fd21d4) |
<a name="5099737a-4001-4518-8990-ea96bb8fb94e"></a>
### 12236-2608762/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:32486
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 93186f612b69:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 1 minutes and 23 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 28 seconds
(8/15) Running minio-dotnet tests ... done in 45 seconds
(9/15) Running minio-go tests ... done in 2 minutes and 34 seconds
(10/15) Running minio-java tests ... done in 2 minutes and 11 seconds
(11/15) Running minio-js tests ... done in 53 seconds
(12/15) Running minio-py tests ... FAILED in 2 minutes and 45 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-203d41d4-a495-4e68-9ce6-b7cd03a8cbcf",
    "object_name": "ed232a13-316b-4e3b-af90-a3c5d7787b38"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-203d41d4-a495-4e68-9ce6-b7cd03a8cbcf, request_id: 167C4916936594F3, host_id: ff54f20c-e8ee-40d3-8048-ecfe2ec257f8, bucket_name: minio-py-test-203d41d4-a495-4e68-9ce6-b7cd03a8cbcf",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1640, in test_thread_safe\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-203d41d4-a495-4e68-9ce6-b7cd03a8cbcf, request_id: 167C4916936594F3, host_id: ff54f20c-e8ee-40d3-8048-ecfe2ec257f8, bucket_name: minio-py-test-203d41d4-a495-4e68-9ce6-b7cd03a8cbcf\n",
  "duration": 6399
}
(12/15) Running s3cmd tests ... done in 19 seconds
(13/15) Running s3select tests ... done in 5 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="2175e2f2-f622-416b-b468-208270fd21d4"></a>
### 12236-2608762/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:32627
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ca4a6aa7f6e2:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... FAILED in 32 seconds
{
  "name": "awscli",
  "duration": 2635,
  "function": "aws --endpoint-url http://minio-dev4.minio.io:32627 s3api copy-object --bucket awscli-mint-test-bucket-28051 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-28051/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 45 seconds
(7/15) Running minio-dotnet tests ... done in 39 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 21 seconds
{
  "args": {},
  "duration": 505,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 2 minutes and 22 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 139,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-sddi8i, objectName = minio-java-test-3i29hh3-copy, resource = /minio-java-test-sddi8i/minio-java-test-3i29hh3-copy, requestId = 167C48DA202EFD5A, hostId = 2a0aae91-4881-4a0c-925c-3f314ef0fbc0)\nrequest={method=PUT, url=http://minio-dev4.minio.io:32627/minio-java-test-sddi8i/minio-java-test-3i29hh3-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-1n5607g/minio-java-test-3i29hh3\nHost: minio-dev4.minio.io:32627\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210505T212747Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210505/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 416\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Wed, 05 May 2021 21:27:47 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 167C48DA202EFD5A\nX-Xss-Protection: 1; mode=block\nDate: Wed, 05 May 2021 21:27:47 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 50 seconds
(9/15) Running minio-py tests ... done in 2 minutes and 44 seconds
(10/15) Running s3cmd tests ... FAILED in 5 seconds
{
  "name": "s3cmd",
  "duration": "3117",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-29372/s3cmd-test-object-24461 9d44a69b-41ee-4240-8857-e8746c818631\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 9d44a69b-41ee-4240-8857-e8746c818631 put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 6 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/iam.go
+++ b/cmd/iam.go
@@ -1114,31 +1114,35 @@ func (sys *IAMSys) NewServiceAccount(ctx context.Context, parentUser string, gro
 	sys.store.lock()
 	defer sys.store.unlock()
 
-	cr, ok := sys.iamUsersMap[parentUser]
-	if !ok {
-		// For LDAP/OpenID users we would need this fallback
-		if sys.usersSysType != MinIOUsersSysType && parentUser != globalActiveCred.ParentUser {
-			_, ok = sys.iamUserPolicyMap[parentUser]
-			if !ok {
-				var found bool
-				for _, group := range groups {
-					_, ok = sys.iamGroupPolicyMap[group]
-					if !ok {
-						continue
-					}
-					found = true
-					break
+	cr, found := sys.iamUsersMap[parentUser]
+	switch {
+	// User found
+	case found:
+		// Disallow service accounts to further create more service accounts.
+		if cr.IsServiceAccount() {
+			return auth.Credentials{}, errIAMActionNotAllowed
+		}
+	// Allow creating service accounts for root user
+	case parentUser == globalActiveCred.AccessKey:
+	// For LDAP/OpenID users we would need this fallback
+	case sys.usersSysType != MinIOUsersSysType:
+		_, ok := sys.iamUserPolicyMap[parentUser]
+		if !ok {
+			var groupHasPolicy bool
+			for _, group := range groups {
+				_, ok = sys.iamGroupPolicyMap[group]
+				if !ok {
+					continue
 				}
-				if !found {
-					return auth.Credentials{}, errNoSuchUser
-				}
+				groupHasPolicy = true
+				break
+			}
+			if !groupHasPolicy {
+				return auth.Credentials{}, errNoSuchUser
 			}
 		}
-	}
-
-	// Disallow service accounts to further create more service accounts.
-	if cr.IsServiceAccount() {
-		return auth.Credentials{}, errIAMActionNotAllowed
+	default:
+		return auth.Credentials{}, errNoSuchUser
 	}
 
 	m := make(map[string]interface{})

